tsdb/head.go

```diff
@@ -524,23 +531,49 @@ Outer:
 					seriesCreationErr = err
 					break Outer
 				}
-
-				if created {
-					// If this series gets a duplicate record, we don't restore its mmapped chunks,
-					// and instead restore everything from WAL records.
-					series.mmappedChunks = mmappedChunks[series.ref]
+				if !created {
+					// There's already a different ref for this series.
+					multiRef[s.Ref] = series.ref
+				}
+				mmc := mmappedChunks[series.ref]
+				if len(series.mmappedChunks) == 0 && len(mmc) > 0 {
+					series.mmappedChunks = mmc
 
 					h.metrics.chunks.Add(float64(len(series.mmappedChunks)))
 					h.metrics.chunksCreated.Add(float64(len(series.mmappedChunks)))
 
 					if len(series.mmappedChunks) > 0 {
 						h.updateMinMaxTime(series.minTime(), series.maxTime())
 					}
-				} else {
-					// TODO(codesome) Discard old samples and mmapped chunks and use mmap chunks for the new series ID.
+				} else if len(mmc) > 0 && len(series.mmappedChunks) > 0 &&
+					series.mmappedChunks[len(series.mmappedChunks)-1].maxTime > mmc[len(mmc)-1].maxTime {
+					// These new m-mapped chunks in the future for the same series.
+					// Hence replace the old m-mapped chunks with new chunks.
+
+					if overlapsClosedInterval(
```

Comment:
Just curious, why would such overlap happen? :thinking:

Reply:
Essentially why would you have multiple `series.mmappedChunks` for the same series.ref?

Reply:
Would be possible if we have the min valid time grace period :)

Reply:
You are right, I should not be using `series.ref` here to get mmap chunks, rather use `s.Ref` from the record. I will fix that in https://github.com/prometheus/prometheus/pull/7438. (So in that case overlapping could be possible for various edge cases)

Reply:
Nice catch about `series.ref`!